src/Groups/matrices/linear_centralizer.jl

```diff
@@ -0,0 +1,426 @@
+function _centralizer_SL(x::MatElem)
+   _,cbm,ED = generalized_jordan_form(x; with_pol=true)    # cbm = change basis matrix
+   n=nrows(x)
+   listgens = MatElem[]
+   _lambda = primitive_element(base_ring(x))
+   res = fmpz(1)
+   ind = fmpz(0)
+   idN = identity_matrix(base_ring(x),n)
+
+   i=1
+   pos=1
+   f=ED[1][1]
+   V=[ED[1][2]]
+   c = evaluate(_centralizer(f), companion_matrix(f))
+# Computes d such that det(c)^d = lambda and replace c by c^d.
+# The result c^d is a primitive root for the base ring of x such that det(c^d) = lambda
+   c = c^(_disc_log(det(c),_lambda))
+# block_dim is a list of [d,m,c], where
+# d = dimension of the Jordan block, m = its multiplicity, c = el of max order determinant in centralizer of the companion matrix
+   block_dim = [[ED[1][2],1,c]]
+   while i <= length(ED)
+      if i<length(ED) && ED[i+1][1]==f
+         i+=1
+         push!(V,ED[i][2])
+         if ED[i][2]==block_dim[length(block_dim)][1] block_dim[length(block_dim)][2] += 1
+         else push!(block_dim, [ED[i][2], 1, c])
+         end
+      else
+         L = _centr_block_unipotent(f,base_ring(x),V; isSL=true)
+         for z in L[1]
+            push!(listgens, insert_block(idN,z,pos,pos))
+         end
+         ind = gcd(ind, gcd([b[1] for b in block_dim]))
+         res *= L[2]
+         pos += degree(f)*sum(V)
+         i+=1
+         if i<=length(ED)
+            f = ED[i][1]
+            V = [ED[i][2]]
+            c = evaluate(_centralizer(f), companion_matrix(f))
+# Computes d such that det(c)^d = lambda and replace c by c^d.
+# The result c^d is a primitive root for the base ring of x such that det(c^d) = lambda
+            c = c^(_disc_log(det(c),_lambda))
+            push!(block_dim, [ED[i][2],1,c])
+         end
+      end
+   end
+
+   # start general blocks, those which have det=1 globally, but not on every single el.div.
+   Ga = abelian_group([Int(order(base_ring(x)))-1 for i in 1:length(block_dim)])
+   Gb = abelian_group(Int(order(base_ring(x)))-1)
+   f = hom(Ga,Gb, [Gb[1]*block_dim[i][1] for i in 1:length(block_dim)])
+   K,g = kernel(f)
+   for k in gens(K)
+   if !iszero(g(k))               #TODO just to do not put identity matrices in the list of generators
+      z = idN
+      pos = 1
+      for i in 1:length(block_dim)
+         insert_block!(z,diagonal_join([block_dim[i][3]^Int(g(k)[i]) for j in 1:block_dim[i][1]]),pos,pos)
+         pos += block_dim[i][1]*block_dim[i][2]*nrows(block_dim[i][3])
+      end
+      push!(listgens, z)
+   end
+   end
+
+   ind = gcd(ind, order(base_ring(x))-1)
+   res = div(res, order(base_ring(x))-1)
+   res *= ind
+   return listgens, res, cbm
+end
+
+
+
+
+########################################################################
+#
+# User level functions
+#
+########################################################################
+
+"""
+    centralizer(G::MatrixGroup{T}, x::MatrixGroupElem{T})
+Return (`C`,`Nothing`), where `C` is the centralizer of `x` in `C`.
```

Comment:
```suggestion

Return (`C`,`nothing`), where `C` is the centralizer of `x` in `C`.
```
But that's not quite accurate, it only returns `nothing` if `G` is GL or SL.

Reply:
Correct it.